File: src/csharp-payloads/templates/recorder-tracks.cs
// Unity eval payload: inspect and remove user-authored RecorderTracks before recording.
//
// The CLI configures Recorder through the RecorderController API inside Play Mode
// (start-recording.cs). A RecorderTrack left on the Timeline records in parallel and
// writes its own files, so the scene must be cleaned before the recording pass is set up.
//
// mode="scan"   read-only; reports the Timeline assets that hold RecorderTracks so the
//               CLI can back them up before anything is mutated.
// mode="remove" deletes those tracks IN MEMORY ONLY. Nothing here saves the scene or the
//               asset, and quit-editor.cs exits without saving, so the .playable files on
//               disk stay untouched and the tracks are back the next time Unity loads them.
var parametersJson = /*__PARAMS_JSON__*/;

string ExtractJsonString(string json, string key)
{
    var match = System.Text.RegularExpressions.Regex.Match(json, "\\\"" + key + "\\\"\\s*:\\s*\\\"((?:\\\\.|[^\\\"\\\\])*)\\\"");
    if (!match.Success)
        throw new System.ArgumentException("Missing required payload parameter: " + key);
    return System.Text.RegularExpressions.Regex.Unescape(match.Groups[1].Value);
}

string EscapeJson(string value)
{
    if (value == null) return "";
    return value.Replace("\\", "\\\\").Replace("\"", "\\\"").Replace("\r", "\\r").Replace("\n", "\\n");
}

string JsonString(string value)
{
    return "\"" + EscapeJson(value) + "\"";
}

string Number(double value)
{
    if (double.IsNaN(value) || double.IsInfinity(value)) return "null";
    return value.ToString("R", System.Globalization.CultureInfo.InvariantCulture);
}

string TrackPath(UnityEngine.Timeline.TrackAsset track)
{
    var parts = new System.Collections.Generic.List<string>();
    var current = track;
    while (current != null)
    {
        parts.Insert(0, current.name);
        current = current.parent as UnityEngine.Timeline.TrackAsset;
    }
    return string.Join("/", parts.ToArray());
}

var directorName = ExtractJsonString(parametersJson, "directorName");
var mode = ExtractJsonString(parametersJson, "mode");
if (mode != "scan" && mode != "remove")
    return "{\"ok\":false,\"error\":\"mode must be scan or remove\"}";

// Resolving the type by name keeps this payload compiling even where Recorder is absent,
// but a silent miss would look like "no tracks to clean" and let a double recording through.
// Not finding the type at all is therefore a hard error, not an empty result.
System.Type recorderTrackType = null;
foreach (var assembly in System.AppDomain.CurrentDomain.GetAssemblies())
{
    var candidate = assembly.GetType("UnityEditor.Recorder.Timeline.RecorderTrack", false);
    if (candidate != null) { recorderTrackType = candidate; break; }
}
if (recorderTrackType == null)
    return "{\"ok\":false,\"error\":\"RecorderTrack type was not found; com.unity.recorder is missing or its API changed\"}";

// setup-recorder.cs selects the director the same way: scan the active scene's roots in
// order. A whole-scene search would also pick up nested directors in an undefined order.
UnityEngine.Playables.PlayableDirector rootDirector = null;
foreach (var root in UnityEngine.SceneManagement.SceneManager.GetActiveScene().GetRootGameObjects())
{
    var candidate = root.GetComponent<UnityEngine.Playables.PlayableDirector>();
    if (candidate != null && candidate.name == directorName)
    {
        rootDirector = candidate;
        break;
    }
}
if (rootDirector == null)
    return "{\"ok\":false,\"error\":\"Root PlayableDirector not found: " + EscapeJson(directorName) + "\"}";

var rootTimeline = rootDirector.playableAsset as UnityEngine.Timeline.TimelineAsset;
if (rootTimeline == null)
    return "{\"ok\":false,\"error\":\"PlayableDirector has no TimelineAsset: " + EscapeJson(directorName) + "\"}";

var warnings = new System.Collections.Generic.List<string>();
var entries = new System.Collections.Generic.List<string>();
var visited = new System.Collections.Generic.HashSet<int>();
var removedCount = 0;

var pending = new System.Collections.Generic.Queue<object[]>();
pending.Enqueue(new object[] { rootTimeline, rootDirector, "root", 0 });

while (pending.Count > 0)
{
    var current = pending.Dequeue();
    var timeline = (UnityEngine.Timeline.TimelineAsset)current[0];
    var owner = (UnityEngine.Playables.PlayableDirector)current[1];
    var chain = (string)current[2];
    var depth = (int)current[3];
    if (timeline == null) continue;
    if (depth > 32)
    {
        warnings.Add("nested-timeline-too-deep: " + chain);
        continue;
    }
    // A ControlTrack can point back at an ancestor director; without this the walk loops.
    if (!visited.Add(timeline.GetInstanceID()))
    {
        warnings.Add("nested-timeline-cycle: " + chain);
        continue;
    }

    // GroupTrack children are not returned by GetOutputTracks in every Timeline version,
    // so walk the tree explicitly: a RecorderTrack inside a group records all the same.
    var flattened = new System.Collections.Generic.List<UnityEngine.Timeline.TrackAsset>();
    var stack = new System.Collections.Generic.Stack<UnityEngine.Timeline.TrackAsset>();
    foreach (var rootTrack in timeline.GetRootTracks()) stack.Push(rootTrack);
    while (stack.Count > 0)
    {
        var track = stack.Pop();
        if (track == null) continue;
        flattened.Add(track);
        foreach (var child in track.GetChildTracks()) stack.Push(child);
    }

    var assetPath = UnityEditor.AssetDatabase.GetAssetPath(timeline);
    var recorderTracks = new System.Collections.Generic.List<UnityEngine.Timeline.TrackAsset>();
    foreach (var track in flattened)
    {
        if (recorderTrackType.IsAssignableFrom(track.GetType()))
        {
            recorderTracks.Add(track);
            continue;
        }

        var controlTrack = track as UnityEngine.Timeline.ControlTrack;
        if (controlTrack == null) continue;
        foreach (var clip in controlTrack.GetClips())
        {
            var control = clip.asset as UnityEngine.Timeline.ControlPlayableAsset;
            if (control == null) continue;
            var target = control.sourceGameObject.Resolve(owner);
            if (target == null)
            {
                warnings.Add("control-clip-unresolved: " + chain + " -> " + TrackPath(controlTrack) + "/" + clip.displayName);
                continue;
            }
            var childDirector = target.GetComponent<UnityEngine.Playables.PlayableDirector>();
            var childTimeline = childDirector == null ? null : childDirector.playableAsset as UnityEngine.Timeline.TimelineAsset;
            if (childDirector == null || childTimeline == null) continue;
            pending.Enqueue(new object[] { childTimeline, childDirector, chain + " > " + clip.displayName, depth + 1 });
        }
    }

    if (recorderTracks.Count == 0) continue;
    if (string.IsNullOrEmpty(assetPath))
        // Without a project path the CLI cannot back the asset up. The removal itself is
        // memory-only, so report it and keep going rather than failing the scene.
        warnings.Add("timeline-asset-path-unavailable: " + chain);

    var trackPaths = new System.Collections.Generic.List<string>();
    foreach (var track in recorderTracks)
    {
        trackPaths.Add(JsonString(TrackPath(track)));
        // Collect first, delete after: DeleteTrack mutates the collections walked above.
        if (mode == "remove")
        {
            timeline.DeleteTrack(track);
            removedCount++;
        }
    }

    entries.Add("{\"assetPath\":" + JsonString(assetPath == null ? "" : assetPath) +
        ",\"chain\":" + JsonString(chain) +
        ",\"tracks\":[" + string.Join(",", trackPaths.ToArray()) + "]}");
}

var warningJson = new System.Collections.Generic.List<string>();
foreach (var warning in warnings) warningJson.Add(JsonString(warning));

// A RecorderTrack clip can be the longest thing on the Timeline, so removing it can
// shorten the duration open-scene.cs reported. Hand the post-removal values back and let
// the CLI re-derive the recording range from them.
return "{\"ok\":true,\"mode\":" + JsonString(mode) +
    ",\"timelines\":[" + string.Join(",", entries.ToArray()) + "]" +
    ",\"removed\":" + removedCount +
    ",\"timelineDurationSec\":" + Number(rootTimeline.duration) +
    ",\"timelineFrameRate\":" + Number(rootTimeline.editorSettings.fps) +
    ",\"warnings\":[" + string.Join(",", warningJson.ToArray()) + "]}";

File: src/batch/recorder-tracks.ts
import { err, ok, type Result } from "../shared/types.js";

/**
 * 録画前に Timeline から取り除く RecorderTrack の扱い。
 *
 * 書き出しの Recorder 構成は Play Mode 内の RecorderController が組み立てる
 * (`start-recording.cs`)。利用者が Timeline へ手で置いた RecorderTrack はそれと
 * 並行して走り、独自の出力先へ書き出してしまうため、録画設定を作る前に外す。
 *
 * 削除は Unity のメモリ上だけで行い、保存はしない(`quit-editor.cs` は保存せず
 * 終了する)。それでも「万一 Editor 側がアセットを保存した」場合に備え、削除前に
 * 対象の Timeline アセットをセッションへバックアップしてから進める。
 */
export interface RecorderTrackTimeline {
	/** プロジェクト相対の Timeline アセットパス。解決できなければ空文字。 */
	readonly assetPath: string;
	/** root からの ControlTrack 経路。ネストのどこで見つかったかを示す。 */
	readonly chain: string;
	readonly tracks: readonly string[];
}

export interface RecorderTrackReport {
	readonly mode: "scan" | "remove";
	readonly timelines: readonly RecorderTrackTimeline[];
	readonly removed: number;
	readonly timelineDurationSec: number | null;
	readonly timelineFrameRate: number | null;
	readonly warnings: readonly string[];
}

export interface RecorderTrackCleanup {
	readonly removed: number;
	readonly warnings: readonly string[];
	/**
	 * 削除後の Timeline 長。RecorderTrack のクリップが Timeline 末尾を伸ばして
	 * いた場合、削除で短くなる。null なら Unity 側が値を返さなかった。
	 */
	readonly timelineDurationSec: number | null;
}

export interface RecorderTrackCleanupDeps {
	readonly evalPayload: (
		mode: "scan" | "remove",
	) => Promise<Result<string, { readonly message: string }>>;
	/**
	 * 削除前に Timeline アセットをバックアップへ登録する。未指定なら保険なしで
	 * 進んだことを警告として残す。
	 */
	readonly registerBackups?: (
		relativePaths: readonly string[],
	) => Promise<Result<void, { readonly message: string }>>;
}

function asStringArray(value: unknown): string[] {
	return Array.isArray(value)
		? value.filter((entry): entry is string => typeof entry === "string")
		: [];
}

function asFiniteNumber(value: unknown): number | null {
	return typeof value === "number" && Number.isFinite(value) ? value : null;
}

export function parseRecorderTrackReport(
	value: string,
): Result<RecorderTrackReport, { readonly message: string }> {
	let parsed: unknown;
	try {
		parsed = JSON.parse(value) as unknown;
	} catch {
		return err({ message: "recorder-tracks payload returned invalid JSON" });
	}
	if (typeof parsed !== "object" || parsed === null)
		return err({ message: "recorder-tracks payload returned invalid JSON" });
	const record = parsed as Record<string, unknown>;
	if (record.ok !== true)
		return err({
			message:
				typeof record.error === "string"
					? record.error
					: "recorder-tracks payload failed",
		});
	const timelines: RecorderTrackTimeline[] = [];
	if (Array.isArray(record.timelines))
		for (const entry of record.timelines) {
			if (typeof entry !== "object" || entry === null) continue;
			const timeline = entry as Record<string, unknown>;
			timelines.push({
				assetPath:
					typeof timeline.assetPath === "string" ? timeline.assetPath : "",
				chain: typeof timeline.chain === "string" ? timeline.chain : "root",
				tracks: asStringArray(timeline.tracks),
			});
		}
	return ok({
		mode: record.mode === "remove" ? "remove" : "scan",
		timelines,
		removed: asFiniteNumber(record.removed) ?? 0,
		timelineDurationSec: asFiniteNumber(record.timelineDurationSec),
		timelineFrameRate: asFiniteNumber(record.timelineFrameRate),
		warnings: asStringArray(record.warnings),
	});
}

function describeTimeline(timeline: RecorderTrackTimeline): string {
	const where =
		timeline.chain === "root"
			? timeline.assetPath || "root Timeline"
			: `${timeline.assetPath || "nested Timeline"} (${timeline.chain})`;
	return `${where}: ${timeline.tracks.join(", ")}`;
}

/**
 * scan で対象を洗い出し、バックアップを取ってから remove を実行する。
 * scan の時点で 0 件なら Unity 側を一切変更しない。
 */
export async function cleanRecorderTracks(
	dependencies: RecorderTrackCleanupDeps,
): Promise<Result<RecorderTrackCleanup, { readonly message: string }>> {
	const scanned = await dependencies.evalPayload("scan");
	if (!scanned.ok) return err({ message: scanned.error.message });
	const scan = parseRecorderTrackReport(scanned.value);
	if (!scan.ok) return scan;

	const warnings = [...scan.value.warnings];
	if (scan.value.timelines.length === 0)
		return ok({
			removed: 0,
			warnings,
			timelineDurationSec: scan.value.timelineDurationSec,
		});

	const assetPaths = scan.value.timelines
		.map((timeline) => timeline.assetPath)
		.filter((assetPath) => assetPath !== "");
	if (dependencies.registerBackups) {
		const registered = await dependencies.registerBackups(assetPaths);
		if (!registered.ok) return err({ message: registered.error.message });
	} else {
		warnings.push(
			"RecorderTrack を削除しましたが、Timeline アセットのバックアップは取得していません。",
		);
	}

	const removedResult = await dependencies.evalPayload("remove");
	if (!removedResult.ok) return err({ message: removedResult.error.message });
	const removed = parseRecorderTrackReport(removedResult.value);
	if (!removed.ok) return removed;

	for (const warning of removed.value.warnings)
		if (!warnings.includes(warning)) warnings.push(warning);
	for (const timeline of removed.value.timelines)
		warnings.push(
			`録画前に RecorderTrack を一時的に外しました — ${describeTimeline(timeline)}`,
		);

	return ok({
		removed: removed.value.removed,
		warnings,
		timelineDurationSec: removed.value.timelineDurationSec,
	});
}
